File: Daily Challenges/2020/July/Pow(x, n).py
class Solution:
    def solve(self, x, n):
        if n == 0:
            return 1
        
        if n == 1:
            return x
        
        if n % 2:
            return x * self.solve(x, n - 1)
        
        temp = self.solve(x, n // 2)
        return temp * temp
    
    def myPow(self, x: float, n: int) -> float:
        ans = self.solve(x, abs(n))
        
        if n < 0:
            return 1.0 / ans
        
        return ans

File: daily_challenge_during_phd/2023/July/Pow(x, n).cpp
class Solution {
public:
    double solve(double x, long long int n) {
        if(n == 0) return 1.0;
        if(n == 1) return x;

        if(n % 2 == 0) {
            double temp = solve(x, n / 2);
            return temp * temp;
        }

        return x * solve(x, n - 1);
    }
    double myPow(double x, int n) {
        if(n < 0) {
            long long int newN = n;
            newN *= -1;

            return 1.0 / solve(x, newN); 
        }

        return solve(x, (long long int) n);
    }
};
